File: src/main/java/wraith/fabricaeexnihilo/client/renderers/BarrelBlockEntityRenderer.java
package wraith.fabricaeexnihilo.client.renderers;

import com.mojang.blaze3d.systems.RenderSystem;
import net.fabricmc.fabric.api.renderer.v1.RendererAccess;
import net.fabricmc.fabric.api.renderer.v1.mesh.MutableQuadView;
import net.fabricmc.fabric.api.transfer.v1.client.fluid.FluidVariantRendering;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidConstants;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.render.RenderLayer;
import net.minecraft.client.render.TexturedRenderLayers;
import net.minecraft.client.render.VertexConsumerProvider;
import net.minecraft.client.render.block.entity.BlockEntityRenderer;
import net.minecraft.client.render.block.entity.BlockEntityRendererFactory;
import net.minecraft.client.render.model.json.ModelTransformationMode;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.item.BlockItem;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.MathHelper;
import net.minecraft.world.World;
import wraith.fabricaeexnihilo.client.BlockModelRendererFlags;
import wraith.fabricaeexnihilo.modules.barrels.BarrelBlockEntity;
import wraith.fabricaeexnihilo.util.Color;

public class BarrelBlockEntityRenderer implements BlockEntityRenderer<BarrelBlockEntity> {

    private static final float XZ_SCALE = 12f / 16f;
    private static final float X_MIN = 2f / 16f;
    private static final float X_MAX = 14f / 16f;
    private static final float Z_MIN = 2f / 16f;
    private static final float Z_MAX = 14f / 16f;
    private static final float Y_MIN = 3f / 16f;
    private static final float Y_MAX = 15f / 16f;

    public BarrelBlockEntityRenderer(BlockEntityRendererFactory.Context ctx) {
    }

    @Override
    public void render(BarrelBlockEntity barrel, float tickDelta, MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, int overlays) {
        var pos = barrel.getPos();
        var world = barrel.getWorld();

        switch (barrel.getState()) {
            case EMPTY -> {}
            case FLUID -> renderFluid(barrel.getFluid(), barrel.getFluidAmount(), matrices, vertexConsumers, light, overlays);
            case ITEM -> renderItem(barrel.getItem(), pos, matrices, vertexConsumers, light, overlays, world);
            case COMPOST -> renderCompost(barrel.getItem(), barrel.getCompostLevel(), barrel.getRecipeProgress(), new Color(0x573f17), pos, matrices, vertexConsumers, light, overlays, world);
        }
    }

    @SuppressWarnings("UnstableApiUsage")
    private void renderFluid(FluidVariant fluid, long amount, MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, int overlays) {
        var sprite = FluidVariantRendering.getSprite(fluid);
        var color = FluidVariantRendering.getColor(fluid);
        var r = ((color >> 16) & 255) / 256f;
        var g = ((color >> 8) & 255) / 256f;
        var b = (color & 255) / 256f;

        if (sprite == null) return;

        RenderSystem.enableDepthTest();

        // Idea stolen from Modern Industrialisation
        var emitter = RendererAccess.INSTANCE.getRenderer().meshBuilder().getEmitter();
        emitter.square(Direction.UP, X_MIN, Z_MIN, X_MAX, Z_MAX, 1 - MathHelper.lerp(amount / (float) FluidConstants.BUCKET, Y_MIN, Y_MAX));
        emitter.spriteBake(sprite, MutableQuadView.BAKE_LOCK_UV);
        vertexConsumers.getBuffer(RenderLayer.getTranslucent()).quad(matrices.peek(), emitter.toBakedQuad(sprite), r, g, b, light, overlays);
    }

    private void renderItem(ItemStack stack, BlockPos pos, MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, int overlays, World world) {
        var yScale = Y_MAX - Y_MIN;

        matrices.push();
        matrices.translate(0.5, Y_MIN + yScale / 2, 0.5);
        matrices.scale(XZ_SCALE, yScale, XZ_SCALE);
        MinecraftClient.getInstance().getItemRenderer().renderItem(stack, ModelTransformationMode.NONE, light, overlays, matrices, vertexConsumers, world, (int) pos.asLong());
        matrices.pop();
    }

    private void renderCompost(ItemStack result, double amount, double progress, Color color, BlockPos pos, MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, int overlays, World world) {
        var finalColor = Color.average(Color.WHITE, color, Math.pow(progress, 4));
        var r = finalColor.r;
        var g = finalColor.g;
        var b = finalColor.b;

        matrices.push();

        var yScale = MathHelper.clamp((float) ((Y_MAX - Y_MIN) * Math.min(amount, 1.0f)), 0, 1);

        if (result.getItem() instanceof BlockItem blockItem) {
            var block = blockItem.getBlock().getDefaultState();
            var model = MinecraftClient.getInstance().getBakedModelManager().getBlockModels().getModel(block);
            var consumer = vertexConsumers.getBuffer(TexturedRenderLayers.getEntityCutout());
            matrices.translate(X_MIN, Y_MIN, X_MIN);
            matrices.scale(XZ_SCALE, yScale, XZ_SCALE);
            BlockModelRendererFlags.setColorOverride(true);
            MinecraftClient.getInstance().getBlockRenderManager().getModelRenderer().render(matrices.peek(), consumer, block, model, r, g, b, light, overlays);
            BlockModelRendererFlags.setColorOverride(false);
        } else {
            matrices.translate(X_MIN, Y_MIN, X_MIN);
            matrices.scale(XZ_SCALE, yScale, XZ_SCALE);
            MinecraftClient.getInstance().getItemRenderer().renderItem(result, ModelTransformationMode.NONE, light, overlays, matrices, vertexConsumers, world, (int) pos.asLong());
        }
        matrices.pop();
    }
}

File: src/main/kotlin/exnihilofabrico/client/renderers/BarrelBlockEntityRenderer.kt
package exnihilofabrico.client.renderers

import alexiil.mc.lib.attributes.fluid.volume.FluidVolume
import exnihilofabrico.modules.barrels.BarrelBlockEntity
import exnihilofabrico.modules.barrels.modes.AlchemyMode
import exnihilofabrico.modules.barrels.modes.CompostMode
import exnihilofabrico.modules.barrels.modes.FluidMode
import exnihilofabrico.modules.barrels.modes.ItemMode
import exnihilofabrico.util.Color
import net.minecraft.client.MinecraftClient
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.render.block.entity.BlockEntityRenderDispatcher
import net.minecraft.client.render.block.entity.BlockEntityRenderer
import net.minecraft.client.render.model.json.ModelTransformation
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import kotlin.math.pow

class BarrelBlockEntityRenderer(dispatcher: BlockEntityRenderDispatcher): BlockEntityRenderer<BarrelBlockEntity>(dispatcher) {


    private val xzScale = 12.0f / 16.0f
    private val xMin = 2.0f / 16.0f
    private val xMax = 14.0f / 16.0f
    private val zMin = 2.0f / 16.0f
    private val zMax = 14.0f / 16.0f
    private val yMin = 0.1875f
    private val yMax = 0.9375f

    override fun render(barrel: BarrelBlockEntity?, tickDelta: Float, matrices: MatrixStack?, vertexConsumers: VertexConsumerProvider?, light: Int, overlays: Int) {
        if(matrices == null || vertexConsumers == null)
            return
        val mode = barrel?.mode ?: return

        when(mode) {
            is FluidMode -> renderFluid(mode, barrel.pos, tickDelta, matrices, vertexConsumers, light, overlays)
            is ItemMode -> renderItemMode(mode, barrel.pos, tickDelta, matrices, vertexConsumers, light, overlays)
            is AlchemyMode -> renderAlchemy(mode, barrel.pos, tickDelta, matrices, vertexConsumers, light, overlays)
            is CompostMode -> renderCompost(mode, barrel.pos, tickDelta, matrices, vertexConsumers, light, overlays)
        }
    }
    fun renderFluid(mode: FluidMode, pos: BlockPos, tickDelta: Float, matrices: MatrixStack, vertexConsumers: VertexConsumerProvider, light: Int, overlays: Int) {

    }
    fun renderItemMode(mode: ItemMode, pos: BlockPos, tickDelta: Float, matrices: MatrixStack, vertexConsumers: VertexConsumerProvider, light: Int, overlays: Int) {
        renderItem(mode.stack, pos, tickDelta, matrices, vertexConsumers, light, overlays, 1.0f)
    }
    fun renderAlchemy(mode: AlchemyMode, pos: BlockPos, tickDelta: Float, matrices: MatrixStack, vertexConsumers: VertexConsumerProvider, light: Int, overlays: Int) {

    }
    fun renderCompost(mode: CompostMode, pos: BlockPos, tickDelta: Float, matrices: MatrixStack, vertexConsumers: VertexConsumerProvider, light: Int, overlays: Int) {

        val yScale = (yMax-yMin) * minOf(mode.amount.toFloat(), 1.0f)

        val color = Color.average(Color.WHITE, mode.color, mode.progress.pow(4))

//        matrices.push()
//        matrices.translate(pos.x+0.5,(pos.y+yMin+yScale/2).toDouble(), pos.z+0.5)
//        matrices.scale(xzScale,yScale,xzScale)
//        RenderHelper.renderItemColored(mode.result, ModelTransformation.NONE, color)
//        matrices.pop()
    }

    fun renderItem(stack: ItemStack, pos: BlockPos, tickDelta: Float, matrices: MatrixStack, vertexConsumers: VertexConsumerProvider, light: Int, overlays: Int, level: Float) {
        val yScale = (yMax-yMin)*level

        matrices.push()
        matrices.translate(pos.x+0.5,(pos.y+yMin+yScale/2).toDouble(), pos.z+0.5)
        matrices.scale(xzScale,yScale,xzScale)
        MinecraftClient.getInstance().itemRenderer.renderItem(stack, ModelTransformation.Mode.NONE, light, overlays, matrices, vertexConsumers)
        matrices.pop()
    }

    @Deprecated("1.14 renderer")
    fun render(barrel: BarrelBlockEntity?, x: Double, y: Double, z: Double, partialTicks: Float, breakingStage: Int) {
        val mode = barrel?.mode ?: return

        when(mode) {
            is FluidMode -> renderFluidMode(mode, x, y, z)
//            is ItemMode -> renderItem(mode.stack, 1.0, x, y, z)
            is AlchemyMode -> renderAlchemyMode(mode, x, y, z)
            is CompostMode -> renderCompostMode(mode, x, y, z)
        }
    }

    private fun renderAlchemyMode(mode: AlchemyMode, x: Double, y: Double, z: Double) {
        when(mode.before) {
            is FluidMode -> renderFluidMode(mode.before, x, y, z)
//            is ItemMode -> renderItem(mode.before.stack, 1.0, x, y, z)
            is AlchemyMode -> renderAlchemyMode(mode.before, x, y, z)
            is CompostMode -> renderCompostMode(mode.before, x, y, z)
        }
        when(mode.after) {
            is FluidMode -> renderFluidMode(mode.after, x, y, z)
//            is ItemMode -> renderItem(mode.after.stack, 1.0, x, y, z)
            is AlchemyMode -> renderAlchemyMode(mode.after, x, y, z)
            is CompostMode -> renderCompostMode(mode.after, x, y, z)
        }
    }

    private fun renderCompostMode(mode: CompostMode, x: Double, y: Double, z: Double) {
        val yScale = (yMax-yMin) * minOf(mode.amount, 1.0)

        val color = Color.average(Color.WHITE, mode.color, mode.progress.pow(4))

//        GlStateManager.pushMatrix()
//        GlStateManager.translated(x+0.5,y+yMin+yScale/2, z+0.5)
//        GlStateManager.scaled(xzScale,yScale,xzScale)
//        GlStateManager.color4f(color.r, color.g, color.b, color.a)
//        RenderHelper.renderItemColored(mode.result, ModelTransformation.Type.NONE, color)
//        GlStateManager.popMatrix()
    }

    private fun renderFluidMode(mode: FluidMode, x: Double, y: Double, z: Double) {
        renderFluidVolume(mode.fluid, mode.fluid.amount.toDouble() / FluidVolume.BUCKET, x, y, z)
    }
    fun renderFluidVolume(volume: FluidVolume, level: Double, x: Double, y: Double, z: Double) {
        val yRender = (yMax-yMin)*level + yMin

//        GuiLighting.disable()
//        volume.render(listOf(FluidRenderFace.createFlatFace(xMin, yMin, zMin, xMax, yRender, zMax, 16.0, Direction.UP)), x, y, z)
//        GuiLighting.enable()
    }
}